packages/dds/shared-object-base/src/sharedObject.ts

```diff
@@ -823,14 +829,26 @@ export interface ISharedObjectKind<TSharedObject> {
 	create(runtime: IFluidDataStoreRuntime, id?: string): TSharedObject;
 }
 
+/**
+ * Defines a kind of shared object.
+ * Used in containers to register a shared object implementation, and to create new instances of a given type of shared object.
+ * @privateRemarks
+ * Type erased reference to an {@link ISharedObjectKind} or a DataObject class.
+ * For use by the declarative API only.
+ * @public
+ */
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface SharedObjectKind<out TSharedObject = unknown>
```

Comment:
This is the important new type in this change. The old ISharedObjectKind is appropriate for the encapsulated API, and this is a type erased version removing its members for use in the declarative API.

Reply:
We try to start all interfaces with "I".
I'm not sure that dropping I makes it clearer on what type actually means / where to use it.

Should there be any limitations on TSharedObject? Like extend ISharedObject?

Reply:
> We try to start all interfaces with "I".
> I'm not sure that dropping I makes it clearer on what type actually means / where to use it.

Generally, we have been naming types with members that define some sort of contract, usually a collection of methods "I" something.
This type is not like that. It has no members and is not implementable. Its an opaque token as far as the user of it is concerned.
It is defined using `interface` because TSC produces better IntelliSense and error messages if we use `interface` then if we use `type` in this case (using interface introduces a name that can occur in messages while `type` does not).
As an implementation detail this type actually extends the ErasedType class (as thats the only want to make a non-implementable type), so its not even a real interface from the class vs interface perspective.
Thus I think not having an "I" is acceptable here.

The version with the "I" is an actual interface which is alpha/legacy, actually legal and safe to implement, and has the API surface needed for the encapsulated API. Only `SharedObjectKind` is public, so public users shouldn't get confused.

Encapsulated API users should stick with the one that actually has members they can access, but getting it wrong isn't very risky as the compiler will let you know and the runtime behavior would work in most cases anyway.

That was my reasoning, but if you have other name suggestions, I'd be happy to consider them. This was just the first option I thought of, and seemed ok and didn't require renaming existing types.

> Should there be any limitations on TSharedObject? Like extend ISharedObject?

I think not, but thats not a strong opinion.

I don't think adding such an extends clause gains us anything (other than perhaps some documentation, which might be worth it). Its mainly helpful if for code works with SharedObjectKinds without using strong typing of its output, which we don't do (since we down cast to internal types first), and if someone did want such code, they can just include the constraint it in their own code. Generally I find it cleaner to leave the constrains out unless we have a reason to include them.

For this specific case if we constrained it to ISharedObject, that would add a bunch of types back to the public API. If we wanted a constraint here IFluidLoadable would be the good choice as thats all ITree extends publicly.
Additionally when we actually stabilize any DataObjects, they can work via this type without us adding any extra API surface (they work correctly with it currently: we just don't have any public non-experimental data objects), so that would also suggest IFluidLoadable if we want a constraint here.

Lastly as this is a type only we implement, adding constraints to is non-breaking, but removing them is breaking, which is another reason to side on leaving it unconstrained for now.